tweetcloud/helpers/text.py

```diff
@@ -0,0 +1,80 @@
+import collections
+import datetime
+import html
+import re
+from itertools import groupby
+from typing import Counter, Dict, Final, Iterable, Optional
+
+from twitter import Status
+from wordcloud import STOPWORDS
+
+stop_words: Final = {"go", "will"}.union(set(STOPWORDS))
+URL_PATTERN: Final = re.compile(r"https?://")
+EMAIL_PATTERN: Final = re.compile(r"[\w.\-+%]+@[\w.\-]+\.[a-zA-Z]{2,}")
+MENTIONS_PATTERN: Final = re.compile(r"@\w+")
+stop_patterns: Final = (URL_PATTERN, MENTIONS_PATTERN, EMAIL_PATTERN)
+BAD_CHAR_PATTERN: Final = re.compile(r"[\W\d]+")
+
+
+def transform_tweets_to_word_counts(
+    tweets: Iterable[Status],
+) -> Dict[datetime.date, Counter[str]]:
+    """
+    Group tweets and return normalized word counts per group.
+    """
+    groups = groupby(sorted(tweets, key=_tweet_group_key), key=_tweet_group_key)
+    # Lots of iteration here. If we start processing tons of tweets
+    # it might be noticeable, but if it comes to that we should use
+    # a proper text processing library.
+    return {
+        key: sum(map(normalize_and_split_tweet, group), start=collections.Counter())
+        for key, group in groups
+    }
+
+
+def _tweet_group_key(tweet: Status) -> datetime.date:
+    """
+    Get the grouping key for each tweet, i.e., the Monday prior to the
```

Comment:
How do you feel about using Sunday before? Since calendars start on Sunday and end on Saturday.

Reply:
sure, but I just updated the docs. The original method was also Monday to Sunday, we were just writing Sunday to Sunday in the wordcloud title. `.weekday` goes from Monday to Sunday (following ISO standards, and I think most of the world) so this is what we did before:

```python
sunday = datetime.date(2020, 7, 27)
weekstart = sunday - datetime.timedelta(days=sunday.weekday() + 1)
assert weekstart == datetime.date(2020, 7, 20)
```

Do we want to explicitly go around it? We could potentially do something like this:
```python
sunday = datetime.date(2020, 7, 27)
# ISO standards say weekday is 1-7 starting Monday, so isoweekday (mod 7) gives 0-6 starting Sunday
weekstart = sunday - datetime.timedelta(days=sunday.isoweekday() % 7)
assert weekstart == datetime.date(2020, 7, 27)
```

Feel like it might be better to stick to standards, but I don't think it's a problem to do it our own way. fwiw almost all weekly groupings I've encountered are either a) start on Monday or b) start on first day of dataset and group 7 days.

Reply:
Ah, I didn't realize the ISO standard was M-S. let's just keep as is then!